### ISSUE TITLE
Failing Test : [sig-api-machinary] Etcd failure [Disruptive]

### ISSUE THREAD

**Issue by mohammedzee1000** [kubernetes/kubernetes#66770]
# Failing Jobs
[sig-release-master-upgrade#gce-1.10-master-upgrade](https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-1.10-master-upgrade-master)

# Failing Tests
[[sig-api-machinery] Etcd failure [Disruptive] should recover from SIGKILL](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1668)

/sig api-machinery
/kind failing-test
/priority important-soon
/milestone v1.12

**Comment by wenjiaswe** [comment]
The purpose of this test is to intentionally kill etcd and see if the system recovers. Before 7/31, this test was flacky, not failing all the time. I looked at [one of failure at 18:47 on 7/31](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1683), apiserver healthz check failed on etcd failure for a couple of times and got killed by kubelet before etcd recovered. Here is the timeline:

- 12:37:16.977392, the last apiserver healthz check failed on etcd, there after, apiserver is shutting down
- 12:37:18.281277, etcd restarted
- 12:37:21.319, test failed. 
 > INFO: Couldn't delete ns: "e2e-tests-etcd-failure-j55l5": Delete https://35.188.145.131/api/v1/namespaces/e2e-tests-etcd-failure-j55l5: dial tcp 35.188.145.131:443: getsockopt: connection refused (&url.Error{Op:"Delete", URL:"https://35.188.145.131/api/v1/namespaces/e2e-tests-etcd-failure-j55l5", Err:(*net.OpError)(0xc424833c70)})
- 12:37:22.468387, apiserver recovered

So basically, etcd recovery is less than 2 seconds too long than the test timeout. The other failures are similar too. So I suspect that the resource is somehow held up by other things at the moment. 

And I realize that after #66730, the test always passes so far. I am not familiar with that fix, @dekkagaijin , do you think if k8s.gcr.io/metadata-proxy:v0.1.10 could have any impact on this?

**Comment by dekkagaijin** [comment]
It'd be very surprising if #66730 was related. That change just updates the GCE metadata proxy to allow for `gcloud` to authenticate itself when metadata concealment is enabled.

**Comment by wenjiaswe** [comment]
@dekkagaijin, thanks for the clarification. @roycaihw , I am remotely working from China this week , it seems like I don't have access to [[sig-api-machinery] Etcd failure [Disruptive] should recover from SIGKILL](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1668) anymore, maybe because it's running on GCE and there is no access to GCE here? Anyway, would you please help to see whether this failure is still a problem? Thanks!

**Comment by wenjiaswe** [comment]
@dekkagaijin, thanks for the clarification. @roycaihw , I am remotely working from China this week , it seems like I don't have access to [[sig-api-machinery] Etcd failure [Disruptive] should recover from SIGKILL](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1668) anymore, maybe because it's running on GCE and there is no access to GCE here? Anyway, would you please help to see whether this failure is still a problem? Thanks!

**Comment by roycaihw** [comment]
The test `[sig-api-machinery] Etcd failure [Disruptive] should recover from SIGKILL` in job https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-1.10-master-upgrade-master failed twice between 8/3 and 8/7 (today):

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1692
```
Started | 2018-08-04 14:45 PDT
Elapsed | 12h38m
```
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1693
```
Started | 2018-08-05 03:24 PDT
Elapsed | 12h25m
```

I'll take a look to see if the two failures' behavior still matches https://github.com/kubernetes/kubernetes/issues/66770#issuecomment-410478558

**Comment by roycaihw** [comment]
etcd is taking >30s to recover, which already exceeded the [sleeping time](https://github.com/kubernetes/kubernetes/blob/a1b3133b4648ab804280a4c26ee1c5015c708d19/test/e2e/apimachinery/etcd_failure.go#L86) (20s) in the [e2e test](https://github.com/kubernetes/kubernetes/blob/a1b3133b4648ab804280a4c26ee1c5015c708d19/test/e2e/apimachinery/etcd_failure.go#L64-L70)

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1692
etcd took 35s to recover
```
(etcd last previous log was at 2018-08-05 05:47:32.045810 I | mvcc: finished scheduled compaction at 76335 (took 2.141101ms))

E2E test killed etcd and then slept for 20s
05:49:54.019] failing etcd

E2E test querying apiserver
05:50:14.311 deleting pods from existing replication controller

the last apiserver healthz check failed before apiserver gets killed
05:50:25.277929       1 wrap.go:42] GET /healthz: (3.442000165s) 500

kubelet decided to kill apiserver
05:50:27.998952    1508 kuberuntime_manager.go:549] Container "kube-apiserver" ({"docker" "b5f2b16d8d798509c1fb04e7631be770e657648733dac943d164455a511bef31"}) of pod kube-apiserver-bootstrap-e2e-master_kube-system(d90dc4fdfd9719be8e1316f25d4a2530): Container failed liveness probe.. Container will be killed and recreated.

etcd restarted
05:50:29.143751 I | pkg/flags: recognized and used environment variable ETCD_SNAPSHOT_COUNT=10000

E2E test failed and teared down test namespace
05:50:35.213] [AfterEach] [sig-api-machinery] Etcd failure [Disruptive]

apiserver restarts
05:50:36.177053       1 flags.go:27] FLAG: --address="127.0.0.1"
```

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1693
etcd took 31s to recover
```
(etcd last previous log was at 2018-08-05 21:41:41.326661 I | mvcc: finished scheduled compaction at 113290 (took 10.066944ms))

E2E test killed etcd and then slept for 20s
21:43:08.435 failing etcd

E2E test querying apiserver
21:43:28.699 deleting pods from existing replication controller

etcd restarted
21:43:39.686431 I | pkg/flags: recognized and used environment variable ETCD_SNAPSHOT_COUNT=10000

the last apiserver healthz check failed before apiserver gets killed
21:43:42.950523       1 wrap.go:42] GET /healthz: (3.820478578s) 500

kubelet decided to kill apiserver
21:43:42.957716    1500 kuberuntime_manager.go:549] Container "kube-apiserver" ({"docker" "a6a95cdd07858c5348bbb511cc26a323abd3d45c9e681b45916700b175320153"}) of pod kube-apiserver-bootstrap-e2e-master_kube-system(d90dc4fdfd9719be8e1316f25d4a2530): Container failed liveness probe.. Container will be killed and recreated.

E2E test failed and teared down test namespace
21:43:46.135] [AfterEach] [sig-api-machinery] Etcd failure [Disruptive]

apiserver restarts
21:43:48.009834       1 flags.go:27] FLAG: --address="127.0.0.1"
```
